src/settings.js

```diff
@@ -936,7 +936,10 @@ var EXPLICIT_ZEXT = 0;
 
 // Global variable to export the module as for environments without a
 // standardized module loading system (e.g. the browser and SM shell).
-var EXPORT_NAME = 'Module';
+// Default EXPORT_NAME is 'Module', but if building directly to .html
+// with -s MODULARIZE=1 but not with -s MODULARIZE_INSTANCE=1, then default EXPORT_NAME
+// is 'EmscriptenCode' to not conflict with the default provided Module object.
+var EXPORT_NAME = '';
```

Comment:
the comment text is not fully clear, and I still don't get it myself maybe - so we have the MODULARIZE function, and the Module object. What do we use the Module object for, that conflicts here? In MODULARIZE I would hope that it is used for nothing, and any existing uses are bugs we should fix? I know you fixed the separate-asm one, are there more?

Reply:
`src/shell.html` and `src/shell_minimal.html` both have

```js
var Module = {
  ...
};
```

and we have tons of code patterns showing `Module` to be a global configuration object that has both the imports and exports. So this puts out an idiom that people will write `var Module = { ... };` configuration objects in their own shells as well.

So then if I do `-s MODULARIZE=1 -o a.html`, I will get `a.js` have `function Module(Module) {`, but my `a.html` also has `var Module = { configs here };`, which contradict each other.

Or if I do `-s MODULARIZE=1 -o a.js` and wrote my own `a.html` to launch it, following tacit practices shown by `src/shell.html` that showcases that one should write a `var Module = { ... };` config object, then again these two conflict, as my `a.html` has a `var Module = { ... };`, but when the compiled `a.js` file is loaded in, it will override it with a `function Module(...) {...}`.

The above aspects can lead to confusion of "what on earth is 'Module'?". Sometimes it is a global config object that has both imports+exports+library_x.js specific config vars. Sometimes it is the name of the JS module function that is generated (as EXPORT_NAME='Module' was default value in -s MODULARIZE=1 mode). In MODULARIZE=1 mode inside the generated module it is again the imports object (i.e. `function {{{EXPORT_NAME}}}(Module) { ... }`). In `MODULARIZE_INSTANCE=1` builds it is first the imports object, but then is replaced by assigning the export object over it.

Requiring the value of EXPORT_NAME= to come directly from the user and never giving it a default value (especially `EXPORT_NAME='Module'`) is a good step towards clearing the confusion, since people will then less easily mix these together.

Reply:
I see, very good points.

I think we should fix this before this PR. What I'd recommend is requiring `EXPORT_NAME` to not be `Module`. This would be a breaking change, so we need to prepare users, but it definitely seems necessary given the above - otherwise we'll be stuck trying to work around this in complicated ways in our codebase forever. (Sad we didn't realize this earlier, but better now than later.)

Specifically, what I'm suggesting is that the default value of `EXPORT_NAME` be something like `EmscriptenCode` as you suggested. (or maybe just `Code`? probably too short) and in emcc.py we assert on it not being set to `Module`, with a clear error.

If you don't have time for this, I could try to find time myself.

Reply:
Reverted away from doing any changes to this, and just focused on the MODULARIZE+USE_PTHREADS combination.